Grammar, content, and type changes (#1)
aoxrud-ww/web-accessibility-tips: README.md

<main class="main-content"></main>
```

### <a name="navigation"></a>Navigation
```html
<!-- bad -->
<div class="navigation">
  <a href="">Link 1</a>
</div>

<!-- good -->
<nav class="sidebar">
  <a href="">Link 1</a>
</nav>

<!-- better -->
<nav class="sidebar">
  <ul aria-label="Main Navigation">
    <li><a href="">Link 1</a></li>
  </ul>
</nav>
```

## <a name="role"></a>The `role` attribute

> Most of the semantic HTML elements can be expressed as non-standard elements (ie. `<div>`, `<span>`, etc...) elements with a specific [`role` attribute](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques#roles).

Reviewer: ```suggestion
Most of the semantic HTML elements can be expressed as non-standard elements (ie. `<div>`, `<span>`, etc...) with a specific [`role` attribute](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques#roles).
```

However, more often than not, screen readers don't handle these "faked" elements the same way. 
These inconsistencies might lead one to add additional functionality (keyboard shortcuts, announcements) to get the same behavior as a semantic element.

Rule of thumb: if there is a semantic element for the purpose, use semantic element.

## <a name="additionalContext"></a>Providing additional context 

There are times where the content and styling around an interactive element provides the context for the label for that element. For example: Consider a page where there are many links that say "View more" and nothing else.

Some ways assistive technologies navigate a page is by reciting content on the page, like all links in a page. 
Screen readers will not describe the styling surrounding the entity they are reading.

> This poses a problem when there are many links that don't have unique labels.

Reviewer: ```suggestion
This poses a problem when there are many links that are visually different, but don't have unique labels.
```

> When screen reader users recite all links on the page, they won't fully understand the context for the view more link. Providing additional context via one of the methods below can help.

Reviewer: ```suggestion
When screen readers recite all links on the page, users won't fully understand the context for the view more link. Providing additional context via one of the methods below can help.
```

}
```

> Why not just use `display: none;`? If you were to do so the element would have it's display turned off. This will make the element and all it's descendants inaccessible to screen readers.

Reviewer: ```suggestion
Why not just use `display: none;`? If you were to do so the element would have it's display turned off. This will make the element and all it's descendants invisible and inaccessible to screen readers.
```

#### First determine if the `alt` attribute should be empty

In some cases the `alt` attribute should be an empty string when the image:

- Is purely decorative. Ex: a button says "settings" and has an image of a gear. The gear is decorative, it provides no additional information.
- Provides no additional information, but is there to convey a mood. Ex: a lifestyle image followed by a form. The lifestyle image, though full of content, provides no meaningful relation or additional context to the main content (ie the form).
- The content of the image is conveyed in the surrounding text. Ex: an image of a cat with a caption underneath that says "cat". Adding alt text stating "cat" would be redundent

```html
<img src="./someDecorativeImg.svg" alt="" />
```

#### Determining the non-empty `alt` attribute value

If the image provides additional context to the page, it needs a thoughful description.

Imagine you have closed your eyes and have never seen the image before:
- How would someone describe the image to you?

> - What the image provides additional context to the page?

Reviewer: ```suggestion
- What context does the image provide to the page?
```

- Avoid using words like: "image", "illustration", "picture", etc... since the user will know they are looking at a picture.

> - Be as succient as possible.

Reviewer: ```suggestion
- Be as succinct as possible.
```

```html
<div style="background-image: url(...)">
  <span class="visually-hidden">description of background image</span>
</div>
```

### <a name="svg"></a>SVG

> When possible SVG images are prefered over bitmap. Some uses use screen enlargers to better view the content; SVG images scale losslessly while bitmap images become pixelated.

Reviewer: ```suggestion
When possible use SVG images over bitmap images because they can be enlarged via page zoom without losing quality; SVG images scale losslessly while bitmap images become pixelated.
```

1. Add a `<title>` node inside of the svg that describes the image.

```html
<svg>
  <title>description of svg</title>
</svg>
```

> 2. Hide the svg via `aria-hidden="true"` and add a visually hidden element that describes it.

Reviewer: In markdown you can prefix each item in list with `1. ` and markdown will take care of enumerating it. This makes it easier to reorder lines without having to manually change the numbers in each line

Author: @aoxrud-ww I'd normally agree with that method but bc of the `html` blocks separating them they were both rendered as `1. `

<img width="494" alt="Screen Shot 2021-03-10 at 4 48 09 PM" src="https://user-images.githubusercontent.com/42351805/110702452-86311680-81c0-11eb-91fe-e906503232f8.png">

```html
<label for="firstname">First name:</label>
<input type="text" name="firstname" id="firstname">
```

Sometimes we don't want to visually display the `<label>` so we can [**visually hide**](#hiding) the element using our previously described `visually-hidden` class:

```html
<label for="firstname" class="visually-hidden">First name:</label>
<input type="text" name="firstname" id="firstname">
```

## <a name="inputValidation"></a>Input Validation

Inputs are often given different styling to denote an invalid input. This styling may not be obvious to assistive technologies. Styling should never be the sole way of denoting an invalid input. Instead styling should be used in conjunction with a text message and the [aria-invalid](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute) attribute.

```html
<input aria-invalid="true" />
```

## <a name="disablingInputs"></a>Disabling Inputs

Assistive technologies wouldn't announce the disabled styling so the disabled state should be indicated using the [aria-disabled](https://www.digitala11y.com/aria-disabled-state/) attribute. 

> There are exceptions where the inputs are styled as visually disabled and they don't have the `disabled` attribute declared. For example: a checkbox needs to be visually disabled but it should still be focusable. We'd want this so its content is still accessible to the screen reader.

Reviewer: Not sure this checkbox example is an actual use-case. We wouldn't denote a checkbox as disabled but still have it checkable. It would be conflicting information to the user.
We did use the pattern where buttons are disabled but still clickable and when you do that it triggers the form validation

Author: @aoxrud-ww I was confused here. I thought aria-disabled hid elements from the screen reader but that's not the case. That being said I think this section could benefit from a brief discussion of `disabled` vs `aria-disabled` 
```suggestion
For example: a button needs to be shown as disabled and it should still be clickable.
Maybe clicking on the disabled button triggers a validation announcement.
It's important to note that the `disabled` attribute will hide the element from the screen reader. This should be avoided because although the element is disabled it still provides valuable information to the user. For example if a next button has a `disabled` (instead of an `aria-disabled`) attribute by default until some condition is met that button will be hidden from the screen reader. Knowing that there **is** a next button and it's disabled is valuable information which can be conveyed through `aria-disabled`.
```

```html
<button aria-disabled="true" class="button--disabled">Save</button>
```

## <a name="contrast"></a>Contrast

> Contrast is an important consideration when styling text and its background. If contrast is too low the text can become illegible. "Contrast" is defined as a measure of the difference in perceived brightness between two colors (text color and background color). Contrast is expressed as a ratio ranging from 1:1 (white on white) to 21:1 (black on a white). To be legible the contrast must be 1:4.5 at a miniumum. Use WebAIM's [contrast checker](https://webaim.org/resources/contrastchecker/) to test text and background color.

Reviewer: ```suggestion
Contrast is an important consideration when styling text and its background. If contrast is too low the text can become illegible. "Contrast" is defined as a measure of the difference in perceived brightness between two colors (text color and background color). Contrast is expressed as a ratio ranging from 1:1 (white on white) to 21:1 (black on a white). To be legible the contrast must be 1:4.5 at a minimum. Use WebAIM's [contrast checker](https://webaim.org/resources/contrastchecker/) to test text and background color.
```

updateResultsCount(10, "apple")
```

**Beware:** if you trigger multiple announcements that contain the same message, some screen readers only announce the first one and ignore subsequent ones.
To avoid this announced-once issue, please ensure that each alert message is different.

**Beware:** the container that has `aria-live` must exist on the DOM before the content within is modified.
If `aria-live` container is added along with new content it will not be picked up by some screen readers.

A related attribute is [aria-atomic](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions#additional_live_region_attributes) which indicates to the screen reader whether it should announce:
- the entire block when anything changes, or
- the individual change by itself

## <a name="expandableContent"></a>Collapsible/Expandable Content

> A common design pattern is to have a button that reveals some section, like an accordion or a modal.

Reviewer: ```suggestion
A common design pattern is to have a button that reveals some section, like an accordion.
```
When opening a modal, the focus should move which ignores this status announcement
